ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCompletion/Rules/NamedUnionCaseFieldsPatRule.fs

```diff
@@ -0,0 +1,98 @@
+namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Features.CodeCompletion.Rules
+
+open FSharp.Compiler.Symbols
+open JetBrains.ReSharper.Feature.Services.CodeCompletion.Infrastructure
+open JetBrains.ReSharper.Feature.Services.CodeCompletion.Infrastructure.AspectLookupItems.BaseInfrastructure
+open JetBrains.ReSharper.Feature.Services.CodeCompletion.Infrastructure.AspectLookupItems.Behaviors
+open JetBrains.ReSharper.Feature.Services.CodeCompletion.Infrastructure.AspectLookupItems.Info
+open JetBrains.ReSharper.Feature.Services.CodeCompletion.Infrastructure.AspectLookupItems.Matchers
+open JetBrains.ReSharper.Feature.Services.CodeCompletion.Infrastructure.AspectLookupItems.Presentations
+open JetBrains.ReSharper.Feature.Services.CodeCompletion.Infrastructure.LookupItems
+open JetBrains.ReSharper.Plugins.FSharp.Psi
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.CodeCompletion
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.CodeCompletion.FSharpCompletionUtil
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Parsing
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open JetBrains.ReSharper.Psi
+open JetBrains.ReSharper.Psi.ExpectedTypes
+open JetBrains.ReSharper.Psi.ExtensionsAPI
+open JetBrains.ReSharper.Psi.Resources
+
+[<Language(typeof<FSharpLanguage>)>]
+type NamedUnionCaseFieldsPatRule() =
+    inherit ItemsProviderOfSpecificContext<FSharpCodeCompletionContext>()
+
+    // Find the parameterOwner `A` in `A()` or `A(a = a; )`
+    // Filter out the already used fields
+    let getFieldsFromParametersOwnerPat (parametersOwnerPat: IParametersOwnerPat) (filterFields: Set<string>) =
+        let referenceName = parametersOwnerPat.ReferenceName
+        if isNull referenceName then Array.empty else
+
+        // We need to figure out if `A` actually is a UnionCase.
+        let fcsUnionCase = referenceName.Reference.GetFcsSymbol().As<FSharpUnionCase>()
+        if isNull fcsUnionCase then Array.empty else
+
+        let fieldNames =
+            fcsUnionCase.Fields
+            |> Seq.choose (fun field -> if field.IsNameGenerated then None else Some field.Name)
+            |> Seq.toArray
+
+        // Only give auto completion to the fields only when all of them are named.
+        if fieldNames.Length <> fcsUnionCase.Fields.Count then Array.empty else
```

Comment:
We should probably allow it for named fields anyway:

<img width="848" alt="Screenshot 2023-06-27 at 18 11 33" src="https://github.com/JetBrains/resharper-fsharp/assets/3923587/1622d30c-bb45-41c6-a2b5-dd37f1f8fb17">